### PR TITLE
Improve type fixing logic

### PR DESCRIPTION
**Pull request by wendigo**
We can update data in place as the Jackson deserializes Lists to ArrayLists

### CODE REVIEW

File: client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java [hunk 1]
```diff
@@ -217,61 +217,19 @@ private static ColumnTypeHandler createTypeHandler(ClientTypeSignature signature
...
                 return new RowClientTypeHandler(signature);
             case BIGINT:
-                return (value) -> {
-                    if (value instanceof String) {
-                        return Long.parseLong((String) value);
-                    }
-                    return ((Number) value).longValue();
-                };
-
+                return BigintTypeHandler.INSTANCE;
```
Comment (losipiuk) [inline client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java:214]: What is the difference? I would expect the lambda is inlined into static code anyway.

File: client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java [hunk 2]
```diff
@@ -68,16 +68,13 @@ public static Iterable<List<Object>> fixData(List<Column> columns, List<List<Obj
...
                 throw new IllegalArgumentException("row/column size mismatch");
             }
-            ArrayList<Object> newRow = new ArrayList<>(typeHandlers.length);
-            int column = 0;
-            for (Object value : row) {
+            for (int i = 0; i < row.size(); i++) {
+                Object value = row.get(i);
                 if (value != null) {
-                    value = typeHandlers[column].fixValue(value);
+                    row.set(i, typeHandlers[i].fixValue(value)); // update value in-place
```
Comment (losipiuk) [inline client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java]: how can you assume here that provided list is modifiable. What is the purpose of this change?

Comment (wendigo) [reply client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java]: it is modifiable since Jackson returns it as ArrayList.

File: client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java (hunk 1 above)
Comment (wendigo) [reply client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java:214]: I don't want JVM spending time trying to figure out whether lambda can be inlined or not

File: client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java [hunk 3]
```diff
@@ -113,15 +110,14 @@ private ArrayClientTypeHandler(ClientTypeSignature signature)
...
         {
-            List<?> listValue = (List<?>) value;
-            ArrayList<Object> fixedValues = new ArrayList<>(listValue.size());
-            for (Object element : listValue) {
+            List<Object> listValue = (List<Object>) value;
+            for (int i = 0; i < listValue.size(); i++) {
+                Object element = listValue.get(i);
                 if (element != null) {
-                    element = elementHandler.fixValue(element);
+                   listValue.set(i, elementHandler.fixValue(element));
```
[the remark this answers is not in the document]
Comment (wendigo) [reply client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java]: As above. ArrayLists

**Comment by wendigo** [comment]
If the assumptions will change in the future we will know as this code will break. For now, it doesn't make sense to copy rows

File: client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java [hunk 4]
```diff
@@ -63,17 +63,19 @@ public static Iterable<List<Object>> fixData(List<Column> columns, List<List<Obj
...
+                    if (value != null) {
+                        row.set(i, typeHandlers[i].fixValue(value)); // update value in-place
+                    }
                 }
+                rows.add(unmodifiableList(row)); // allow nulls in list
             }
-            rows.add(unmodifiableList(row)); // allow nulls in list
+        }
+        catch (IndexOutOfBoundsException e) {
+            throw new IllegalArgumentException("row/column size mismatch");
```
Comment (losipiuk) [inline client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java]: This will only catch the case when we have for objects in row than types. Not the other way around.

File: client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java [hunk 5]
```diff
@@ -56,28 +56,29 @@ final class FixJsonDataUtils
...
-    public static Iterable<List<Object>> fixData(List<Column> columns, List<List<Object>> data)
+    public static Iterable<List<Object>> fixData(List<Column> columns, List<List<Object>> rows)
     {
-        if (data == null) {
+        if (rows == null) {
             return null;
         }
         ColumnTypeHandler[] typeHandlers = createTypeHandlers(columns);
-        ImmutableList.Builder<List<Object>> rows = ImmutableList.builderWithExpectedSize(data.size());
+        // Jackson deserializes lists to ArrayList's so instead of copying lists, update them in place
```
Comment (losipiuk) [inline client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java:65]: this is fragile

File: client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java [hunk 6]
```diff
@@ -56,28 +56,29 @@ final class FixJsonDataUtils
 {
     private FixJsonDataUtils() {}
 
-    public static Iterable<List<Object>> fixData(List<Column> columns, List<List<Object>> data)
+    public static Iterable<List<Object>> fixData(List<Column> columns, List<List<Object>> rows)
     {
-        if (data == null) {
+        if (rows == null) {
```
Comment (losipiuk) [inline client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java:61]: you are changing this code multiple times in different commits. Hard to read. Please extract meaningless changes (renames etc) as preparatory commits and make one commit which changes logic here.

File: client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java (hunk 1 above)
Comment (losipiuk) [reply client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java:214]: You should. It is JVMs job. If it you do not know it is explicitly needed, prooved by benchmarks please revert.

File: client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java (hunk 2 above)
Comment (losipiuk) [reply client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java]: The fact that you can does not mean you should. It is fragile. What are we gaining.

File: client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java (hunk 5 above)
Comment (wendigo) [reply client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java:65]: but well tested - if this changes in some future jackson release - we will know ;)

**Comment by losipiuk** [comment]
I do not like this change to be honest. Unless we know it is important perf-wise I would drop.
